server/channels/app/users/password.go

```diff
@@ -4,28 +4,41 @@
 package users
 
 import (
-	"errors"
 	"strings"
 
-	"golang.org/x/crypto/bcrypt"
-
 	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/v8/channels/app/password/hashers"
 )
 
 func CheckUserPassword(user *model.User, password string) error {
-	if err := ComparePassword(user.Password, password); err != nil {
-		return NewErrInvalidPassword("")
+	if password == "" || user.Password == "" {
+		return InvalidPasswordError
+	}
+
+	// Validate the stored hash is compliant with the currently used password
+	// hasher, and fail early if it is not
+	hasher, phc := hashers.GetHasherFromPHCString(user.Password)
+	if hasher != hashers.LatestHasher {
+		return OutdatedPasswordHashingError
+	}
+
+	// Run the actual comparison
+	if err := hashers.LatestHasher.CompareHashAndPassword(phc, password); err != nil {
```

Comment:
nit: do we need to expose LatestHasher? Wondering if we can just have a `CompareHashAndPassword` at the package level that uses a private `latestHasher.CompareHashAndPassword` instead?

Reply:
Unexported: https://github.com/mattermost/mattermost/pull/33830/commits/652777fa276c7eb69362ec0b5439e6cefed54036